pkg/agent/controller/networkpolicy/fqdn.go

```diff
@@ -405,80 +414,109 @@ func (f *fqdnController) deleteRuleSelectedPods(ruleID string) error {
+		}
+	}
 
 	f.fqdnSelectorMutex.Lock()
 	defer f.fqdnSelectorMutex.Unlock()
-	oldDNSMeta, exist := f.dnsEntryCache[fqdn]
+	cachedDNSMeta, exist := f.dnsEntryCache[fqdn]
 	if exist {
-		mustCacheResponse = true
-		for ipStr := range responseIPs {
-			if _, ok := oldDNSMeta.responseIPs[ipStr]; !ok {
+		// check for new IPs.
+		for newIPStr, newIPMeta := range newIPsWithExpiration {
+			if _, exist := cachedDNSMeta.responseIPs[newIPStr]; !exist {
+				updateIPWithExpiration(newIPStr, newIPMeta)
 				addressUpdate = true
-				break
 			}
 		}
-		for oldIPStr, oldIP := range oldDNSMeta.responseIPs {
-			if _, ok := responseIPs[oldIPStr]; !ok {
-				if oldDNSMeta.expirationTime.Before(time.Now()) {
-					// This IP entry has already expired and not seen in the latest DNS response.
-					// It should be removed from the cache.
+
+		// check for presence of already cached IPs in the new response.
+		for cachedIPStr, cachedIPMeta := range cachedDNSMeta.responseIPs {
+			if newIPMeta, exist := newIPsWithExpiration[cachedIPStr]; !exist {
+				// The IP was not found in current response.
+				if cachedIPMeta.expirationTime.Before(currentTime) {
+					// this IP is expired and stale, remove it by not including it but also signal an update to syncRules.
 					addressUpdate = true
 				} else {
-					// Add the unexpired IP entry to responseIP and update the lowest applicable TTL if needed.
-					responseIPs[oldIPStr] = oldIP
-					if oldDNSMeta.expirationTime.Before(recordTTL) {
-						recordTTL = oldDNSMeta.expirationTime
-					}
+					// It hasn't expired yet, so just retain it with its existing expirationTime.
+					updateIPWithExpiration(cachedIPStr, cachedIPMeta)
 				}
+			} else {
+				// This already cached IP is part of the current response, so update it with max time between received time and its old cached time.
+				expTime := laterOf(newIPMeta.expirationTime, cachedIPMeta.expirationTime)
+				updateIPWithExpiration(cachedIPStr, ipWithExpiration{
+					ip:             cachedIPMeta.ip,
+					expirationTime: expTime,
+				})
 			}
 		}
+
 	} else {
+		// First time seeing this domain.
+		// check if this needs to be tracked, by checking if it matches any Antrea FQDN policy selectors.
+
+		// iterate over current rules mapping
+		addToCache := false
 		for selectorItem := range f.selectorItemToRuleIDs {
 			// Only track the FQDN if there is at least one fqdnSelectorItem matching it.
 			if selectorItem.matches(fqdn) {
-				mustCacheResponse, addressUpdate = true, true
+				// A FQDN can have multiple selectorItems mapped, hence we do not break the loop upon a match, but
+				// keep iterating to create mapping of multiple selectorItems against same FQDN.
+				addToCache = true
 				f.setFQDNMatchSelector(fqdn, selectorItem)
 			}
 		}
+		if addToCache {
+			for ipStr, ipMeta := range newIPsWithExpiration {
+				updateIPWithExpiration(ipStr, ipMeta)
+			}
+			addressUpdate = true
+		}
 	}
-	if mustCacheResponse {
+
+	if len(ipWithExpirationMap) > 0 {
```

Comment:
I would add a comment saying that the only way this condition can be false is if the FQDN doesn't match any selector. Based on my understanding of the code I don't see any other case where that could happen. Am I missing something @Dyanngg ?

Reply:
"or the IPs are not in response and have expired" can you give an example?
I cannot think of a case where the FQDN matches at last one selector, yet `ipWithExpirationMap` is empty. The function returns early if `responseIPs` is empty, and as long as `responseIPs` is non empty and the FQDN matches a selector, can `ipWithExpirationMap` be empty? We will always take all IPs from `responseIPs` and add them to `ipWithExpirationMap` (regardless of expiration time actually, even though presumably it will always be in the future).

Reply:
to clarify, this is just the comment that I think is wrong here, not the implementation